test/proxy/ERC1967/ERC1967Utils.test.js

```diff
@@ -0,0 +1,186 @@
+          expect(web3.utils.hexToNumber(slotValue)).to.equal(newVal);
+        });
+      });
+    });
+  });
+
+  describe('ADMIN_SLOT', function () {
+    beforeEach('set admin', async function () {
+      await setSlot(this.utils, AdminSlot, admin);
+    });
+
+    describe('getAdmin', function () {
+      it('returns current admin and matches admin slot value', async function () {
+        expect(await this.utils.$getAdmin()).to.equal(admin);
+        expect(await getAddressInSlot(this.utils.address, AdminSlot)).to.equal(admin);
+      });
+    });
+
+    describe('changeAdmin', function () {
+      it('sets admin in storage and emits event', async function () {
+        const newAdmin = anotherAccount;
+        const receipt = await this.utils.$changeAdmin(newAdmin);
+
+        expect(await getAddressInSlot(this.utils.address, AdminSlot)).to.equal(newAdmin);
+        expectEvent(receipt, 'AdminChanged', { previousAdmin: admin, newAdmin: newAdmin });
+      });
+
+      it('reverts when setting the address zero as admin', async function () {
+        await expectRevertCustomError(this.utils.$changeAdmin(ZERO_ADDRESS), 'ERC1967InvalidAdmin', [ZERO_ADDRESS]);
+      });
+    });
+  });
+
+  describe('BEACON_SLOT', function () {
+    beforeEach('set beacon', async function () {
+      this.beacon = await UpgradeableBeacon.new(this.v1.address, admin);
+      await setSlot(this.utils, BeaconSlot, this.beacon.address);
+    });
+
+    describe('getBeacon', function () {
+      it('returns current beacon and matches beacon slot value', async function () {
+        expect(await this.utils.$getBeacon()).to.equal(this.beacon.address);
+        expect(await getAddressInSlot(this.utils.address, BeaconSlot)).to.equal(this.beacon.address);
+      });
+    });
+
+    describe('upgradeBeaconToAndCall', function () {
+      it('sets beacon in storage and emits event', async function () {
+        const newBeacon = await UpgradeableBeacon.new(this.v2.address, admin);
+        const receipt = await this.utils.$upgradeBeaconToAndCall(newBeacon.address, EMPTY_DATA);
+
+        expect(await getAddressInSlot(this.utils.address, BeaconSlot)).to.equal(newBeacon.address);
+        expectEvent(receipt, 'BeaconUpgraded', { beacon: newBeacon.address });
+      });
+
+      it('reverts when beacon does not contain code', async function () {
+        await expectRevertCustomError(
+          this.utils.$upgradeBeaconToAndCall(anotherAccount, EMPTY_DATA),
+          'ERC1967InvalidBeacon',
+          [anotherAccount],
+        );
+      });
+
+      it("reverts when beacon's implementation does not contain code", async function () {
+        const newBeacon = await UpgradeableBeacon.new(this.v2.address, admin);
+
+        // Corrupt beacon's implementation
+        // Note that the beacon itself is not ERC1967 compliant, so implementation is at slot 1
+        // after the owner in slot 0.
+        await setSlot(newBeacon, 1, anotherAccount);
```

Comment:
Let's add `expect(await newBeacon().implementation()).to.equal(zero)` here to validate the assumption that the slot is correct.

Reply:
I think you meant something like this:

```js
expect(await newBeacon.implementation()).to.equal(await getAddressInSlot(newBeacon.address, 1));
```

Isn't it?

Reply:
Closing this, see my other comment.

Reply:
Yeah that's what I meant though.

Reply:
Even this is solved I would consider adding it

Reply:
It shouldn't be necessary if we use a purpose-built mock as in my other comment but I'll let you make the decision.